lib/reversetunnel/agent_test.go

```diff
@@ -269,34 +277,37 @@ func TestAgentStart(t *testing.T) {
 		// global requests during startup.
 		<-waitForVersion
 
-		openChannels++
+		atomic.AddInt32(openChannels, 1)
 		assert.Equal(t, name, chanHeartbeat, "Unexpected channel opened during startup.")
 		return &mockSSHChannel{MockSendRequest: func(name string, wantReply bool, payload []byte) (bool, error) {
-			sentPings++
+			atomic.AddInt32(sentPings, 1)
 
 			assert.Equal(t, name, "ping", "Unexpected request name.")
 			assert.False(t, wantReply, "Expected no reply wanted.")
 			return true, nil
 		}}, make(<-chan *ssh.Request), nil
 	}
 
-	versionReplies := 0
 	client.MockReply = func(r *ssh.Request, b1 bool, b2 []byte) error {
+		atomic.AddInt32(versionReplies, 1)
+
 		// Unblock once we receive a version reply.
 		close(waitForVersion)
-		versionReplies++
 
 		assert.Equal(t, versionRequest, r.Type, "Unexpected request type.")
 		assert.Equal(t, teleport.Version, string(b2), "Unexpected version.")
 		return nil
 	}
 
-	err := agent.Start(context.Background())
+	ctx, cancel := context.WithCancel(context.Background())
+	t.Cleanup(cancel)
+
+	err := agent.Start(ctx)
 
 	require.NoError(t, err)
-	require.Equal(t, 1, openChannels, "Expected only heartbeat channel to be opened.")
-	require.GreaterOrEqual(t, 1, sentPings, "Expected at least 1 ping to be sent.")
-	require.Equal(t, 1, versionReplies, "Expected 1 version reply.")
+	require.Equal(t, 1, int(atomic.LoadInt32(openChannels)), "Expected only heartbeat channel to be opened.")
+	require.GreaterOrEqual(t, 1, int(atomic.LoadInt32(sentPings)), "Expected at least 1 ping to be sent.")
+	require.Equal(t, 1, int(atomic.LoadInt32(versionReplies)), "Expected 1 version reply.")
```

Comment:
are these guaranteed to be incremented before `agent.Start` returns, or does that happen in the background?

Reply:
It is expected to happen before `agent.Start` returns.